ros/src/twist_controller/twist_controller_mod.py

```diff
@@ -32,60 +32,55 @@ def __init__(self, default_update_interval, wheel_base, steer_ratio, min_speed,
-                      .format(velo_p, accel_p, accel_i))
+    def setup_pid_controllers(self, velo_p, velo_i, braking_p, braking_i):
+        rospy.loginfo("Initializing PID controllers with velo_P: {}, velo_I: {}, braking_P: {}, braking_I: {}"
+                      .format(velo_p, velo_i, braking_p, braking_i))
         # create velocity pid controller thresholded between min and max
         # acceleration values
         self.velocity_pid_controller = PID(
-            velo_p, 0, 0, max_deceleration, max_acceleration)
+            velo_p, velo_i, 0, 0, 1)
 
         # create acceleration pid controller thresholded between 0% and 100%
         # for throttle
-        self.acceleration_pid_controller = PID(accel_p, accel_i, 0.0, 0.0, 1.0)
+        self.braking_pid_controller = PID(
+            braking_p, braking_i, 0.0, 0.0, 10000)
 
     def control(self, target_linear_velocity, target_angular_velocity, current_linear_velocity):
         # compute timestep
         timestep = self.compute_timestep()
-        # use velocity pid controller to compute desired acceleration
-        acceleration_command = self.compute_acceleration_command(
-            target_linear_velocity - current_linear_velocity, timestep)
-        rospy.logwarn('Current linear velocity %.2f, target linear velocity %.2f, desired total acceleration %.2f',
-                      current_linear_velocity, target_linear_velocity, acceleration_command)
+        velocity_error = target_linear_velocity - current_linear_velocity
+
+        if (target_angular_velocity == 0 and current_linear_velocity == 0):
+            # reset integrators if we're at a stop
+            self.reset()
+
+        # use throttle if we want to speed up or if we want to slow down just slightly
+        throttle_command = self.velocity_pid_controller.step(
+            velocity_error, timestep) if velocity_error > 0 or velocity_error > (-0.075 * current_linear_velocity) else 0
+
+        # use brake if we want to slow down somewhat significantly or it looks like want to stop
+        brake_command = self.braking_pid_controller.step(-velocity_error, timestep) if velocity_error < (-0.1 * current_linear_velocity) or (velocity_error < 0 and current_linear_velocity < 2.5) else 0
+
+        rospy.logdebug('Current linear velocity %.2f, target linear velocity %.2f, throttle_command %.2f, brake_command %.2f',
+                      current_linear_velocity, target_linear_velocity, throttle_command, brake_command)
 
         # Return throttle, brake, steer
-        # TODO - build "coasting" to reduce speed rather than always using positive throttle or brake
-        return (self.compute_throttle(acceleration_command, target_linear_velocity, timestep) if acceleration_command > 0. else 0.,
-                self.compute_brake_torque(abs(acceleration_command)) if acceleration_command < 0. else 0.,
+        return (throttle_command,
+                brake_command,
```

Comment:
For issue #50, I think after we are happy with the basic throttle and brake control and rough tuning, we should still apply a final rate limiting or time-constant filtering on each command before they are output.  This would just be to prevent any wild step changes in the output and we would tune the rates to be fast enough to not interfere by too much delay on the control feedback loops, but slow enough to guarantee the output can't spike erratically in case of some glitch or unexpected issue.  For drive-by-wire systems, the actuators probably have their own rate limits anyways, but it's always good to put limits on the control output even just to match how fast a human could realistically slam or release the pedals just in case.

Reply:
Alright. Do you think the LPF that I tried earlier would make sense here? It's primitive, but that seems like that might fit the bill.

Reply:
Yea the LPF would work but tuning by time constants is a little more abstract than a straight linear rate limit.  Since we already have the LPF ready to use, let’s try that first.